File: shared_resources/src/main/java/com/thewizrd/shared_resources/utils/CustomJsonConverter.java
package com.thewizrd.shared_resources.utils;

import com.google.gson.TypeAdapter;
import com.google.gson.stream.JsonReader;
import com.google.gson.stream.JsonWriter;

import java.io.IOException;

public class CustomJsonConverter<T extends CustomJsonObject> extends TypeAdapter<T> {
    private final Class<T> clazz;

    public CustomJsonConverter(Class<T> clazz) {
        this.clazz = clazz;
    }

    @Override
    public void write(JsonWriter out, T value) throws IOException {
        value.toJson(out);
    }

    @Override
    public T read(JsonReader in) {
        try {
            T object = clazz.newInstance();
            object.fromJson(in);
        } catch (InstantiationException | IllegalAccessException e) {
            e.printStackTrace();
        }
        return null;
    }
}

File: shared_resources/src/main/java/com/thewizrd/shared_resources/json/CustomJsonConverter.kt
package com.thewizrd.shared_resources.json

import com.squareup.moshi.JsonAdapter
import com.squareup.moshi.JsonReader
import com.squareup.moshi.JsonWriter
import com.thewizrd.shared_resources.utils.CustomJsonObject
import java.io.IOException

class CustomJsonConverter<T : CustomJsonObject>(private val clazz: Class<T>) : JsonAdapter<T>() {
    @Synchronized
    @Throws(IOException::class)
    override fun fromJson(reader: JsonReader): T? {
        try {
            val `object` = clazz.newInstance()
            `object`.fromJson(reader)
            return `object`
        } catch (e: InstantiationException) {
            e.printStackTrace()
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
        }

        return null
    }

    @Synchronized
    @Throws(IOException::class)
    override fun toJson(writer: JsonWriter, value: T?) {
        if (value == null) {
            writer.nullValue()
        } else {
            value.toJson(writer)
        }
    }
}